### PR TITLE
extend cuda graph size for H200

### PR DESCRIPTION
**Pull request by kushanam**
FILL IN THE PR DESCRIPTION HERE

Adding batch size 512, 768 to the CUDA graph capture list. This will enable efficient use of kv cache for models of size <= llama 70B. This is specifically enabled with H200 cap in mind.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by mgoin — COMMENTED** [review]
Shouldn't the batch size ceiling be set by `max_num_seqs` or `max_num_batched_tokens`? cc @rkooo567 
https://github.com/vllm-project/vllm/blob/6fc4e6e07a55559c3744212b4d562e20d024e661/vllm/engine/arg_utils.py#L89

**Comment by kushanam** [comment]
This shouldn't affect the decisions downstream. max_num_seqs's default is 256 but if increased, without this PR, CUDA graph will not be used unless we expand the range. It comes at the cost of having extra graphs in memory and that is why it is guarded by the available memory and SM version.

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -62,8 +62,11 @@
 # Capture graphs for token size 1, 2, 4, 8, 16, 24, 32, 40, ..., 256.
 # NOTE: _get_graph_batch_size needs to be updated if this list is changed.
 _BATCH_SIZES_TO_CAPTURE = [1, 2, 4] + [
-    _BATCH_SIZE_ALIGNMENT * i for i in range(1, 33)
+    _BATCH_SIZE_ALIGNMENT * i for i in range(1, 34)
```
Comment (youkaichao) [inline vllm/worker/model_runner.py:65]: can we directly extend this to `128`, i.e. the max batchsize to capture becomes 1024?

Comment (kushanam) [reply vllm/worker/model_runner.py:65]: You want to excused 768? Is there any perf implications for padding 768 up to 1024 when the max seq=768?

Comment (youkaichao) [reply vllm/worker/model_runner.py]: we already use memory pool sharing for cudagraph, and the cost should only scale with the max batchsize.

if you capture batchsize 768, it has the same cost of capturing 760, 752,... together.

File: vllm/worker/model_runner.py [hunk 2]
```diff
@@ -62,7 +62,7 @@
 # Capture graphs for token size 1, 2, 4, 8, 16, 24, 32, 40, ..., 256.
```
Comment (youkaichao) [inline vllm/worker/model_runner.py:62]: also change this line please, 256 --> 1024

Comment (kushanam) [reply vllm/worker/model_runner.py]: Done.

**Comment by rkooo567** [comment]
@mgoin  I think the max cuda graph batch size technically cannot be larger than `max_num_batched_tokens `. 

Btw, can we just do this when we know H200 is used? Increasing the cuda graph bath size 4x requires 4x memory for cuda graph capture (and it is much slower to capture graph).

**Comment by youkaichao** [comment]
@rkooo567  I think this is just the sizes we **want** to capture. The real sizes we capture should be capped by max num seqs.

**Comment by rkooo567** [comment]
hmm but max_num_seqs is hardcoded to be 256 iirc. doesn't this mean it will regress for existing users who don't use h200?

**Comment by youkaichao** [comment]
`max_num_seqs` should be a command line flag we can set.

the max size of graph to capture, is minimum of `max(_BATCH_SIZES_TO_CAPTURE)` and `max_num_seqs` .

increasing `_BATCH_SIZES_TO_CAPTURE` should not hurt existing users.

**Comment by rkooo567** [comment]
ah I see. so because max num seq is 256 hard coded by default, it should not affect the existing users and only affect users who increase it manually. that makes sense

**Comment by youkaichao** [comment]
we do have hard-coded `max(_BATCH_SIZES_TO_CAPTURE)` in the code, which will be affected if we directly change `_BATCH_SIZES_TO_CAPTURE` .

I replaced them with `self.max_batchsize_to_capture`, with is minimum of `self.scheduler_config.max_num_seqs` and `max(_BATCH_SIZES_TO_CAPTURE)`.

this way, it should be sound to add more sizes in `_BATCH_SIZES_TO_CAPTURE` .

**Comment by tjohnson31415** [comment]
The changes in this PR also fix a bug that I was looking that would cause a crash:
https://github.com/vllm-project/vllm/pull/7999
(the fix is ensuring that `_use_captured_graph()` checks against the actual max captured size and not just `_BATCH_SIZES_TO_CAPTURE[-1]`)

So just to link them, this PR will also resolve: #6306 and #7907

**Review by njhill — APPROVED** [review]
I guess the only consideration here is that many more graphs will be captured now for folks who had a larger max_num_seqs configured - which will mean higher startup time and memory use.

**Comment by youkaichao** [comment]
> I guess the only consideration here is that many more graphs will be captured now for folks who had a larger max_num_seqs configured - which will mean higher startup time and memory use.

but they will get better throughput, which is the primary goal if one wants to increase `max_num_seqs` lol

**Pull request merged**